### PR TITLE
VM Debugging

### PR DESCRIPTION
**Pull request by holgerd77**
I am so tired of VM debugging where it becomes extremely hard to follow on the control flow that I would want to add some debugging analog to what is done in the `devp2p` and `client` packages. This should make it significantly easier for us to track bugs like in #1076.

This is a first test commit to gather some feedback, I would continue with this tomorrow if there is no significant objection or critics.

This produces output like the following if activated with e.g. `DEBUG=vm:tx ts-node test.ts`:

image

(so only the upper part, JSON below is just a `console.log()` from the script run)

### CODE REVIEW

**Comment by cgewecke** [comment]
100% in favor 🙂

**Comment by holgerd77** [comment]
Have added some first debug loggers `vm:tx`, `vm:tx:gas` and `vm:state` which can be run all at once with `DEBUG=vm:*,vm:tx:gas ts-node test.ts` and I can assure you this will be game-changing for both debugging and educational purposes! 😀  😁  😀

Here is a sample run:

Bildschirmfoto 2021-02-03 um 11 33 51

**Comment by holgerd77** [comment]
Ok, have now also adding a debug logger for opcode based logging, this comes in really really handy, e.g. by just logging on all the `SSTORE` processings with `DEBUG=vm:tx,vm:evm,vm:ops:sstore,vm:*:gas ts-node test.ts` //cc @jochem-brouwer: 😄 

image

**Comment by holgerd77** [comment]
Ok, I'll stop here, have also added some additional README instructions, this is now ready for review. 😄 

@jochem-brouwer I'll try to add an additional `vm:block` logger during next week if you think this is a useful addition which should be prioritized since it will get useful for our client runs.

Please note though that this is meant to be a general purpose debugging helper not necessarily (only) meant to be for our client runs but for all sorts of development and user activity (adding new opcodes, testing some functionality changes in the `evm.ts` code, debugging and/or understanding tx runs,...). Should be very usable for our client testing as well though, as loggers can be arbitrarily combined as necessary. 😁  I also very much assume that we will add additional loggers over time.

**Comment by jochem-brouwer** [comment]
I have a few suggestions after just reading the readme: (can be in a subsequent PR)

- It is very helpful if we know what the current address is we are running operations on. Just the depth is not enough. If we SSTORE, we want to know on which address we are storing. (Same with CREATE)
- It is also very helpful to know if a `CALL*` / `CREATE*` succeeds or not. So, if the depth changes and gets lower (so a call returns), then we should check the top of the stack. If this is nonzero, the call/create has succeeded. If this is not the case, then the call/create has failed. (This thus also means that any storage writes / code deposits did not succeed and are reverted). This is somewhat tricky, since if you call an address which has no code, we don't change the depth (it immediately succeeds, and is usually just used to transfer ether).

In general what would also help is to verify our `step` event against the data which is reported if we call `debug_traceTransaction` on geth. Then we can figure out exactly at which step our results diverge from that of Geth.

**Comment by holgerd77** [comment]
Hi Jochem, thanks for the suggestions, just an answer on the first one on a first round:

If I get this right the operations address on a `CALL` is the `to` address and on `CREATE` the created address, right? Both are available in the `SSTORE`-(or other bytecode)-surrounding log messages. Correct me if I am wrong?

So e.g.

```shell
vm:evm New message caller=0x147013436bd5c7def49a8e27c7fba8ac2b9dfe1f gasLimit=276872 to=0x9ca228250f9d8f86c23690074c2b96d5f5479f79 value=0 delegatecall=no +0ms
...and then closing with
vm:evm message checkpoint +96ms
```

And for CREATE

```shell
vm:evm New message caller=0x9ca228250f9d8f86c23690074c2b96d5f5479f79 gasLimit=243805 to= value=0 delegatecall=no +0ms
  vm:evm Message CREATE execution (to undefined) +0ms
  vm:evm Reduce sender (0x9ca228250f9d8f86c23690074c2b96d5f5479f79) balance (-> 40000000000000000) +1ms
  vm:evm Generated CREATE contract address 0x8398ff6c618e9515468c1c4b198d53666cbe8462 +1ms
... and closing again with
vm:evm message checkpoint committed +0ms#
```

**Comment by holgerd77** [comment]
Isn't point 2 (CALL / CREATE succeeds or not) included in the message results? 🤔

So e.g.:

```shell
vm:evm Received message results gasUsed=106040 execResult: [ gasUsed=106040 exceptionError= returnValue=0000000000000000000000008398ff6c618e9515468c1c4b19... gasRefund=undefined ] +11ms
```

File: packages/vm/lib/evm/interpreter.ts [hunk 1]
```diff
@@ -193,6 +197,29 @@ export default class Interpreter {
...
+      gas: '0x' + eventObj.gasLeft.toString('hex'),
+      gasCost: '0x' + eventObj.opcode.fee.toString(16),
+      stack: hexStack,
+      depth: eventObj.depth,
+    }
+
+    if (!(nameLC in this.opDebuggers)) {
+      this.opDebuggers[nameLC] = createDebugLogger(`vm:ops:${nameLC}`)
+    }
+    this.opDebuggers[nameLC](JSON.stringify(opTrace))
```
Comment (cgewecke) [inline packages/vm/lib/evm/interpreter.ts:221]: This is really nice.

@jochem-brouwer You mentioned wanting to diff the geth / ethereumjs traces - would the most convenient thing be to have two files with JSONs of the respective traces? 

If so debug has [custom formatters][1] which might help with this. 

I think you could collect the steps into a trace and then write to `ejs.vm.<txhash>.json`, when `DEBUG=vm:savetrace` (or something)

[1]: https://www.npmjs.com/package/debug#custom-formatters

File: packages/vm/lib/evm/evm.ts [hunk 2]
```diff
@@ -297,12 +331,15 @@ export default class EVM {
...
+        debug(`Not enough gas or code size not allowed (>= Homestead)`)
         result = { ...result, ...OOGResult(message.gasLimit) }
       } else {
         // we are in Frontier
+        debug(`Not enough gas or code size not allowed (Frontier)`)
         if (totalGas.sub(returnFee).lte(message.gasLimit)) {
           // we cannot pay the code deposit fee (but the deposit code actually did run)
-          result = { ...result, ...COOGResult(totalGas.sub(returnFee)) }
+          //result = { ...result, ...COOGResult(totalGas.sub(returnFee)) }
+          result.gasUsed = totalGas.sub(returnFee)
```
Comment (cgewecke) [inline packages/vm/lib/evm/evm.ts:342]: Out of curiosity - this is a simplification e.g the result is already formatted as COOGResult? 

Should [COOGResult][1] be removed in this PR?

[1]: https://github.com/ethereumjs/ethereumjs-monorepo/blob/80093e6f650427d3cf25d15f2b12fd7e3948c803/packages/vm/lib/evm/evm.ts#L78-L92

**Review by cgewecke — APPROVED** [review]
Looks great to me! 💯

File: packages/vm/lib/evm/evm.ts (hunk 2 above)
Comment (holgerd77) [reply packages/vm/lib/evm/evm.ts:342]: Oh oh oh, this actually sneaked in along doing debugging work for #1076, this shouldn't be here. 😏

@jochem-brouwer just to make you aware, can you please take care of this along working on #1081 respectively along some rebase since you are touching this code anyhow? Thanks!

File: packages/vm/lib/evm/interpreter.ts (hunk 1 above)
Comment (holgerd77) [reply packages/vm/lib/evm/interpreter.ts:221]: I would guess this is a good suggestion since comparing might be easiest on a per file basis with some linux CLI tools? @jochem-brouwer WDYT?
